### DISCUSSION TITLE
Deploying in Heroku

### DISCUSSION THREAD

**Discussion by bevsxyz** [umputun/remark42#1067] [Q&A]
Hi,
I had tried to deploy remark42 with docker to Heroku, but I couldn't get it working. Are there any resources on how to do it? Has anyone got it working successfully?

I'll get back with more details on my attempt.

**Comment by umputun** [comment]
pls use discussions for questions like this. Issues are for the actual issues and remark42 has nothing special that makes it non-deployable to a particular platform.

_Moved to dicussion_

**Reply by bevsxyz** [reply]
Thank you for correcting me!

**Comment by paskal** [comment]
@bevsxyz could you please describe what you do, what result you expect, and what do you get instead? Also, if you're following some guide or documentation, please link it.

**Comment by Ma-Pe-Ma** [comment]
Hello!

Before jumping in this topic I have to clarify that I'm an absolute beginner in this field so sorry for my shortcomings in advance.

I successfully composed the docker image [with the help of the manual](https://remark42.com/docs/getting-started/installation/) and it's perfectly running locally on my machine.

I followed this [guide](https://devcenter.heroku.com/articles/container-registry-and-runtime#building-and-pushing-image-s): I tagged, pushed and released the image successfully to my Heroku account. (I did not forget to modify the compose-file.)

But when the app launches the initialization file fails:
* ERROR: /srv/var doesn't exist, which means that state of the application
* ERROR: will be lost on container stop or restart.
* ERROR: Please mount local directory to /srv/var in order for it to work.
* /srv/init.sh failed
* Process exited with status 1
* State changed from starting to crashed

(I acquired these messages with heroku logs --tail)

And I'm stuck here right now. I don't know if need to:
* compose the docker image differently
* or launch the Heroku app differently (probably with proper arguments?)

I understand that the problem is with storing the application data but I have no idea where to start to solve this problem.

Note: Heroku provides an ["ephemeral"](https://devcenter.heroku.com/articles/active-storage-on-heroku#ephemeral-disk) storage.
Does this affect hosting Remark42 on Heroku? I think this complicates further the problem.

**Reply by paskal** [reply]
Give me some time, I'll try to make it work and will write a guide on that.

**Reply by Ma-Pe-Ma** [reply]
Thanks! Don't hurry. It's not urgent.

Just a side note: I read a little more about it. If I'm not mistaken I have to mount a storage with Heroku at the required path (as noted by the error message).

But this is not so easy because you have to use a third party service like Amazon S3 (as explained in the [aforementioned](https://devcenter.heroku.com/articles/active-storage-on-heroku#ephemeral-disk) storage documentation).

But correct me, if I'm writing nonsense stuff.

**Reply by paskal** [reply]
I've got some time for it now. I'll report back once I'll have the results.

**Reply by paskal** [reply]
I wrote instructions on how to do it, and then I hit the same issue you were asking us about: it's [impossible](https://devcenter.heroku.com/articles/dynos#ephemeral-filesystem) to persistently store data inside the Docker container.

Here are the types of storage available:
- Data (comments, users): `bolt` or `rpc`.
- Admin: `shared` or `rpc`.
- Avatar: `fs` or `bolt`. And also, `uri` which is only supported by `server` but not for avatar migration command and means "fs, bolt or MongoDB".
- Image: `fs`, `bolt` or `rpc`.

Effectively that means that there is no possibility right now to make Remark42 work in Heroku. The only way forward is to create some non-local storage for images (like Amazon S3 you mentioned) and return Mongo support to the Data storage (like discussed in #955).

**Reply by paskal** [reply]
@Ma-Pe-Ma, unfortunately, I think we won't support this setup in the foreseeable future. I'll try to find someplace with a free tier and write doc on it so it would be easier for people to start using remark42 if they don't have their server yet and might use some help getting one.

**Reply by Ma-Pe-Ma** [reply]
Thanks for looking after this problem and confirming that it can't be used on Heroku!

That would be indeed great if it was possible to host it on some free service. I want to use Remark42 in simple projects (for a few users on closed sites and periodically for limited times) and I think it would be a huge overkill to pay/subscribe for some advanced cloud hosting service.

**Reply by Ma-Pe-Ma** [reply]
As I see, 'not paying for it' is the bottleneck for this problem.

Is it possible to host RM42 on a trusted service which costs only a few bucks (<5$) per month?

**Comment by paskal** [comment]
I wrote an instruction on [how to host Remark42 yourself](https://terrty.net/2022/self-hosted-comments-remark42/) in Oracle cloud (for free) or Digital Ocean ($4-6 a month), that's the most I can do given that it's impossible to host it on Heroku.
